src/os/bluestore/BlueStore.cc

```diff
@@ -4819,15 +4841,16 @@ void BlueStore::_txc_finish_kv(TransContext *txc)
     txc->onreadable_sync = NULL;
   }
   if (txc->onreadable) {
-    finisher.queue(txc->onreadable);
+    finishers[txc->osr->parent->shard_hint.hash_to_shard(m_finisher_num)]->queue(txc->onreadable);
     txc->onreadable = NULL;
   }
   if (txc->oncommit) {
-    finisher.queue(txc->oncommit);
+    finishers[txc->osr->parent->shard_hint.hash_to_shard(m_finisher_num)]->queue(txc->oncommit);
     txc->oncommit = NULL;
   }
   while (!txc->oncommits.empty()) {
-    finisher.queue(txc->oncommits.front());
+    auto f = txc->oncommits.front();
+    finishers[txc->osr->parent->shard_hint.hash_to_shard(m_finisher_num)]->queue(f);
```

Comment:
unsigned n = txc->osr->parent->shard_hint.hash_to_shard(m_finisher_num); and then use it for all of these?